### ISSUE TITLE
updates the file then deletes the file

### ISSUE THREAD

**Issue by frakier** [superzanti/ServerSync#86]
#### Serversync Version: 
serversync-2.6.15.jar
#### Minecraft Version: 
1.10.2   forge 1.10.2-12.18.3.2281
#### Issue: 
updates the file then deletes the file
<details>
<summary>Logs (serverysync & minecraft)</summary>
<pre>
Sucessfully updated: morefuelsmod-1.10.x-1.6.0.jar
Sucessfully updated: advancedRocketry.cfg
Sucessfully updated: endercore.cfg
Sucessfully updated: EnderIO.cfg
Sucessfully updated: Guide-API.cfg

<>morefuelsmod-1.10.x-1.6.0.jar deleted
<>advancedRocketry.cfg deleted
<>endercore.cfg deleted
<>EnderIO.cfg deleted
<>Guide-API.cfg deleted
</pre>
</details>
[serversyncconfig.zip](https://github.com/superzanti/ServerSync/files/1259272/serversyncconfig.zip)

**Comment by rheimus** [comment]
@frakier Could I persuade you to test out the 2.6.16.aplha.2 version in the releases section of this repository.

I have not been able to reproduce this issue.

**Comment by frakier** [comment]
I should have the serve to myself this weekend I'll try to give it a test and report back.

**Comment by frakier** [comment]
Testing 2.6.16.aplha.2

Updating and then deleting the jar file...
I closed the app and reran it each time...
(First run)
Set everything up [made sure the morefuelsmod] was present and gave it a try...
	-Successfully updated: mods\morefuelsmod-1.10.x-1.6.0.jar
	...
	<>morefuelsmod-1.10.x-1.6.0.jar deleted
(Second run)
Ran the sync again [no problems but at this point the mod did not exist]
	Successfully updated: mods\morefuelsmod-1.10.x-1.6.0.jar
(Third run)
Ran the sync again
	Successfully updated: mods\morefuelsmod-1.10.x-1.6.0.jar
	...
	<>morefuelsmod-1.10.x-1.6.0.jar deleted
Ran the sync again [no problems but at this point the mod did not exist]
	Successfully updated: mods\morefuelsmod-1.10.x-1.6.0.jar
	
morefuelsmod-1.10.x-1.6.0.jar is till a problem but whatever you did fixed the other files which were
all cfg files.

Hope this helps let me know if you need more tests. Will also be submitting some other things I noticed while testing as issues.

**Comment by frakier** [comment]
Tired a test...
the  mcmod.info for morefuelsmod, the "version": "", was empty changed it to "version": "1.6.2",
and reran the sync , still updates and deletes the file every other time.

Got you code in eclipse, going to look it over next week, not a great programmer myself but maybe I will spot something. Until then instructed the players to check the log and run twice if needed.

**Comment by rheimus** [comment]
Most likely culprits are probably:
- [SyncFile.java:populateModInformation](https://github.com/superzanti/ServerSync/blob/master/src/main/java/com/superzanti/serversync/util/SyncFile.java#L134)
- [SyncFile.java:equals](https://github.com/superzanti/ServerSync/blob/master/src/main/java/com/superzanti/serversync/util/SyncFile.java#L237)

Code still has a fair amount of cleanup needed in style & location of files etc, hopefully its small enough to not be too overwhelming though :smile:.

**Comment by P3rf3ctXZer0** [comment]
It seems to happen with certain pcs and a large number of updates.

**Comment by rheimus** [comment]
Fascinating, I really should set up some auto testing for the project 😄 .

Thanks for the info, gives me some starting points to investigate.

**Comment by P3rf3ctXZer0** [comment]
You know I am always glad to help :)

**Comment by DustRaven** [comment]
I have tried to debug the problem by looking at the equals() function, specifically [this](https://github.com/superzanti/ServerSync/blob/4389cd8915a321957b0a613d0a8bf4ceaddce25b/src/main/java/com/superzanti/serversync/util/SyncFile.java#L254) line, which leads to the deletion every other time. I tried to look at the way the names are compared, and it seems as though every single filename is compared to every single filename (mod1:mod1, mod1:mod2, mod1:mod3, mod2:mod1, mod2:mod2, mod2:mod3 and so on). I'm sure this leads to problems on servers with many mods. Since i have no idea how the memory stuff is handled i can only guess, but could it be that the mod list is truncated somehow?

**Comment by rheimus** [comment]
@DustRaven 
Hmm, nice find.

It's quite possible that during the parsing of the mcmod file that color confuses the system. In which case some sanitizing could help, or if it's some specific standard we could find / make a parser for it.

At the moment the files are just being parsed as JSON.

**Comment by rheimus** [comment]
I have added in a specific check, for now, that looks for these color formatting characters at the beginning of versions or names.

If it finds one it removes the character and the following character, all I could find about these seems to suggest that there is only one character after the identifier.

I was highly tempted to just enforce a reasonable versioning scheme by stripping anything that's not [0-9\.\-A-Za-z].

Considering mod authors can use any old string in these fields though it's much safer to just specifically search for what we know is causing issues, also the above search would leave behind the number used to indicate which color is desired. In theory that would not be a problem as both versions in the comparison would have the extra digit, it's not very clean though and would no doubt cause issues down the line eventually.

**Comment by rheimus** [comment]
Testing required: 2.6.17
- Modify a mcmod.info file to use a special character in the version and or name string
- Test both modified and unmodified sync

**Comment by DustRaven** [comment]
Nope, still happening. The mod in question is [Mr. Crayfishs Furniture Mod](https://minecraft.curseforge.com/projects/mrcrayfish-furniture-mod/files/2221679), maybe there is something else off with the mod?

**Comment by rheimus** [comment]
Hmm, I'll try to do some debugging over the weekend.

**Comment by rheimus** [comment]
I think I'll take a different tack on this one.

SyncFiles are set up with a hash of the file they represent when created, there's no particular reason not to use said hash for mod comparison rather than trying to play with user registered version strings.

I switched the hash comparison to be the only method of file comparison and removed the mcmod.info check entirely.

This does appear to solve the issue on my local test environment.

I have updated the attached build on 2.6.17 with this commit.

**Comment by DustRaven** [comment]
Yeah, it works. The mod is no longer being deleted! Thank you very much!

**Comment by P3rf3ctXZer0** [comment]
@rheimus can I have a copy of the newer serversync.

**Comment by rheimus** [comment]
@P3rf3ctXZer0 
There is a built jar attached to the latest release here: https://github.com/superzanti/ServerSync/releases

I'll to some final testing then put a release out on curse as well.
